mne/decoding/time_gen.py

```diff
@@ -1375,3 +1446,71 @@ def _clean_times(self):
             self.y_pred_ = [y_pred[0] for y_pred in self.y_pred_]
         if hasattr(self, 'scores_'):
             self.scores_ = [score[0] for score in self.scores_]
+
+
+def _chunk_data(X, slices, gat, n_orig_epochs, test_epochs):
+    """Smart chunking to avoid memory overload.
+
+    The parallelization is performed across time samples. To avoid overheads,
+    the X data is splitted into large chunks of different time sizes. To
+    avoid duplicating the memory load to each job, we only pass the time
+    samples that are required by each job. The indices of the training times
+    must be adjusted accordingly.
+    """
+
+    # from object array to list
+    slices = [sl for sl in slices if len(sl)]
+    selected_times = np.hstack([np.ravel(sl) for sl in slices])
+    start = np.min(selected_times)
+    stop = np.max(selected_times) + 1
+    slices_ = [sl - start for sl in slices]
+    X_ = X[:, :, start:stop]
+    return (X_, gat.estimators_, gat._cv_splits, slices_,
+            gat.predict_mode, gat.predict_method, n_orig_epochs,
+            test_epochs)
+
+
+def _set_cv(cv, clf=None, X=None, y=None):
+    from sklearn.base import is_classifier
+
+    # Set the default cross-validation depending on whether clf is classifier
+    # or regressor.
+    if check_version('sklearn', '0.18'):
+        from sklearn.model_selection import (check_cv, StratifiedKFold, KFold)
+        if isinstance(cv, (int, np.int)):
+            XFold = StratifiedKFold if is_classifier(clf) else KFold
+            cv = XFold(n_folds=cv)
+        cv = check_cv(cv=cv, y=y, classifier=is_classifier(clf))
+    else:
+        from sklearn.cross_validation import (check_cv, StratifiedKFold, KFold)
+        if isinstance(cv, (int, np.int)):
+            if is_classifier(clf):
+                cv = StratifiedKFold(y=y, n_folds=cv)
+            else:
+                cv = KFold(n=len(y), n_folds=cv)
+
+    # If cv is only defined in terms of n_folds
+    if isinstance(cv, (int, np.int)):
+        # Automatically chose StratifiedKFold if classification else KFold
+        if check_version('sklearn', '0.18'):
+            XFold = StratifiedKFold if is_classifier(clf) else KFold
+            cv = XFold(n_folds=cv)
+        else:
+            if is_classifier(clf):
+                cv = StratifiedKFold(y=y, n_folds=cv)
+            else:
+                cv = KFold(n=len(y), n_folds=cv)
+        cv = check_cv(cv=cv, X=X, y=y, classifier=is_classifier(clf))
+
+    # Extract train and test set to retrieve them at predict time
+    if hasattr(cv, 'split'):
+        cv_splits = [(train, test) for train, test in
+                     cv.split(X=np.zeros_like(y), y=y)]
+    else:
+        # XXX support sklearn.cross_validation cv
+        cv_splits = [(train, test) for train, test in cv]
```

Comment:
no, user may have sklearn > .18 yet, use `sklearn.cross_validation` objects that don't have `split()`